xarray/core/dataarray.py

```diff
@@ -259,7 +259,10 @@ def __init__(
         else:
             # try to fill in arguments from data if they weren't supplied
             if coords is None:
-                coords = getattr(data, 'coords', None)
+
+                if isinstance(data, DataWithCoords):
```

Comment:
`DataWithCoords` is only used for Dataset and DataArray. But putting a `Dataset` into the `DataArray` constructor doesn't do anything useful, so let's just stick with `DataArray`: 
```suggestion
                if isinstance(data, DataArray):
                    coords = data.coords
```

Example of what passing in a Dataset currently gives you:
```
In [25]: xr.DataArray(xr.Dataset(coords={'x': 1}))
Out[25]:
<xarray.DataArray ()>
array(<bound method Mapping.values of <xarray.Dataset>
Dimensions:  ()
Coordinates:
    x        int64 1
Data variables:
    *empty*>, dtype=object)
Coordinates:
    x        int64 1
```
Yes, `coords` is copied, but the `data` attribute is the bound `values()` method from the Dataset object!